### PR TITLE
V1 pool oracle

### PR DESCRIPTION
**Pull request by rahne-sinclair**

### CODE REVIEW

File: solidity/contracts/utility/Oracle.sol
Comment (ororopickpocket) [inline solidity/contracts/utility/Oracle.sol:82]: ema = k * current price + (1-k) * last ema

File: solidity/contracts/utility/Oracle.sol
Comment (ororopickpocket) [inline solidity/contracts/utility/Oracle.sol:79]: first emas start with the current price

File: solidity/contracts/utility/Oracle.sol
Comment (ororopickpocket) [inline solidity/contracts/utility/Oracle.sol:88]: cumulative price is price * time in seconds since last update

File: solidity/contracts/utility/Oracle.sol
Comment (ororopickpocket) [inline solidity/contracts/utility/Oracle.sol:86]: no need to convert to uint64 anymore. we're not storing these values in a struct anymore. it will be cheaper for us to have them uint256 (because we save the conversions).

File: solidity/contracts/utility/Oracle.sol
Comment (ororopickpocket) [inline solidity/contracts/utility/Oracle.sol:75]: are you here just assuming that each asset has 18 decimals?

File: solidity/contracts/utility/Oracle.sol
Comment (ororopickpocket) [reply solidity/contracts/utility/Oracle.sol:82]: also, 1 - k will not work, because we're working with uints.
you should make k a percentage with 2 decimals. so, 1% is 100 and 100% is 10000.
then, you get

```
ema = (k * price + (10000 - k) * ema)/10000
```

Please also use safeMath for all calculations

**Review by ororopickpocket — CHANGES_REQUESTED** [review]

File: solidity/utils/test_deployment_oracle_v1_pool.js
Comment (ororopickpocket) [inline solidity/utils/test_deployment_oracle_v1_pool.js:161]: why?

File: solidity/test/Oracle.js
Comment (ororopickpocket) [inline solidity/test/Oracle.js:75]: how is an update with the same price info a good test?
we should check that the pricing is updated accordingly if the prices change.

File: solidity/contracts/utility/EnumerableSet.sol
Comment (rahne-sinclair) [reply solidity/contracts/utility/EnumerableSet.sol:20]: No, should I remove it?

File: solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol
Comment (rahne-sinclair) [reply solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol:46]: Okay I will make the change in contract and tests

File: solidity/test/Oracle.js
Comment (rahne-sinclair) [reply solidity/test/Oracle.js]: Added the tests for all the price values from this sheet https://docs.google.com/spreadsheets/d/1iQ9u0AxQO5tNnXBcyWoa2YZvEPpmNssn8918Je7uiEY/edit#gid=0
and checked if the EMA is as expected.

File: solidity/utils/test_deployment_oracle_v1_pool.js
Comment (rahne-sinclair) [reply solidity/utils/test_deployment_oracle_v1_pool.js:161]: This script was created to deploy the Liquidity converted v1 pool with Oracle. Have I done something incorrect or missed anything?

File: solidity/contracts/utility/Oracle.sol
Comment (rahne-sinclair) [reply solidity/contracts/utility/Oracle.sol]: Done

File: solidity/contracts/utility/Oracle.sol
Comment (rahne-sinclair) [reply solidity/contracts/utility/Oracle.sol]: Yes, I have updated it in contract and tests

File: solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol
Comment (rahne-sinclair) [reply solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol:508]: Added

File: solidity/utils/test_deployment_oracle_v1_pool.js
Comment (rahne-sinclair) [reply solidity/utils/test_deployment_oracle_v1_pool.js:161]: Do you mean this script not needed and I should add the oracle deployment to the test_deployment_rsk.js script?

Comment (ororopickpocket) [reply solidity/utils/test_deployment_oracle_v1_pool.js:161]: i mean why would you register all of these addresses again although they already have been registered on deployment?

File: solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol
Comment (ororopickpocket) [inline solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol:17]: Better move this to the converterBase. Will make it easier for us to not mess up existing structures with future updates.

**Review by ororopickpocket — APPROVED** [review]

**Comment by ororopickpocket** [comment]
approved with one comment.
please do not merge before we got a review from @eMarchenko

**Comment by ororopickpocket** [comment]
would also be nice if 
1. the oracle deployment would be added to the addConverter script
2. if it would be added to the script for upgrading converters - because that's what we'll need to do.

File: solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol
Comment (rahne-sinclair) [reply solidity/contracts/converter/types/liquidity-pool-v1/LiquidityPoolV1Converter.sol:17]: Ok

File: solidity/utils/test_deployment_oracle_v1_pool.js
Comment (rahne-sinclair) [reply solidity/utils/test_deployment_oracle_v1_pool.js:161]: Oh okay, removing that part.

**Review by korepkorep — COMMENTED** [review]
Several code quality issues that don't endanger contracts` security

**Review by eMarchenko — APPROVED** [review]
@korepkorep has performed [the review](https://github.com/DistributedCollective/oracle-based-amm/pull/31#pullrequestreview-680844540),  no security issues have been discovered

**Pull request merged**
